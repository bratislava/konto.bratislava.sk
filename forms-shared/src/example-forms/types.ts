import { GenericObjectType } from '@rjsf/utils'
import { FormsBackendFile } from '../form-files/serverFilesTypes'
import { ClientFileInfo } from '../form-files/fileStatus'

export type ExampleForm<FormData = GenericObjectType> = {
  name: string
  formData: FormData
  serverFiles?: FormsBackendFile[]
  clientFiles?: ClientFileInfo[]
  sharepointFieldMap?: Record<string, string>
}
