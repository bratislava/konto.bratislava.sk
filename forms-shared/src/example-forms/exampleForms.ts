import priznanieKDaniZNehnutelnostiExample1 from './examples/priznanieKDaniZNehnutelnostiExample1'
import priznanieKDaniZNehnutelnostiExample5 from './examples/priznanieKDaniZNehnutelnostiExample5'
import priznanieKDaniZNehnutelnostiExample2 from './examples/priznanieKDaniZNehnutelnostiExample2'
import priznanieKDaniZNehnutelnostiExample3 from './examples/priznanieKDaniZNehnutelnostiExample3'
import priznanieKDaniZNehnutelnostiExample4 from './examples/priznanieKDaniZNehnutelnostiExample4'
import predzahradkyExample from './examples/predzahradkyExample'
import { ExampleForm } from './types'
import zavazneStanoviskoKInvesticnejCinnostiExample from './examples/zavazneStanoviskoKInvesticnejCinnostiExample'
import komunitneZahradyExample from './examples/komunitneZahradyExample'
import stanoviskoKInvesticnemuZameruExample from './examples/stanoviskoKInvesticnemuZameruExample'

export const exampleForms: Record<string, ExampleForm[]> = {
  'stanovisko-k-investicnemu-zameru': [stanoviskoKInvesticnemuZameruExample],
  'zavazne-stanovisko-k-investicnej-cinnosti': [zavazneStanoviskoKInvesticnejCinnostiExample],
  predzahradky: [predzahradkyExample],
  'komunitne-zahrady': [komunitneZahradyExample],
  'priznanie-k-dani-z-nehnutelnosti': [
    priznanieKDaniZNehnutelnostiExample1,
    priznanieKDaniZNehnutelnostiExample2,
    priznanieKDaniZNehnutelnostiExample3,
    priznanieKDaniZNehnutelnostiExample4,
    priznanieKDaniZNehnutelnostiExample5,
  ],
}

export const exampleDevForms: Record<string, ExampleForm[]> = {
  'ziadost-o-pridelenie-najomneho-bytu': [],
}
