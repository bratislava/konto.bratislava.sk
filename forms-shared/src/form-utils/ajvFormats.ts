import { validate as validateUuid, version as uuidVersion } from 'uuid'
import { electronicFormatIBAN, validateIBAN } from 'ibantools'
// https://stackoverflow.com/a/51177696
export const baTimeRegex = /^(\d|0\d|1\d|2[0-3]):[0-5]\d$/

/**
 * Compares two time strings in format HH:MM.
 *
 * Copied from: https://github.com/ajv-validator/ajv-formats/blob/4dd65447575b35d0187c6b125383366969e6267e/src/formats.ts#L180C1-L186C2
 */
function compareBaTime(s1: string, s2: string): number | undefined {
  if (!(s1 && s2)) return undefined
  const t1 = new Date(`2020-01-01T${s1}`).valueOf()
  const t2 = new Date(`2020-01-01T${s2}`).valueOf()
  if (!(t1 && t2)) return undefined
  return t1 - t2
}

// https://blog.kevinchisholm.com/javascript/javascript-e164-phone-number-validation/
export const baPhoneNumberRegex = /^\+[1-9]\d{10,14}$/

// https://github.com/mnestorov/regex-patterns?tab=readme-ov-file#slovakia
// But `libphonenumber` contains stricter validation rules:
// https://github.com/google/libphonenumber/blob/e3dbc4bde172a49d06e86b8305a7ca624e95cae8/resources/PhoneNumberMetadata.xml#L27018
export const baSlovakPhoneNumberRegex = /^\+421[1-9][0-9]{8}$/

export const parseRatio = (value: string) => {
  const ratioRegex = /^(0|[1-9]\d*)\/([1-9]\d*)$/
  if (!ratioRegex.test(value)) {
    return { isValid: false }
  }

  const parts = value.split('/')
  const numerator = parseInt(parts[0], 10)
  const denominator = parseInt(parts[1], 10)

  if (numerator > denominator) {
    return { isValid: false }
  }

  return { isValid: true, numerator, denominator }
}

export const validateBaFileUuid = (value: unknown): value is string => {
  return typeof value === 'string' && validateUuid(value) && uuidVersion(value) === 4
}

const baAjvInputFormats = {
  'ba-slovak-zip': /\b\d{5}\b/,
  'ba-phone-number': baPhoneNumberRegex,
  'ba-slovak-phone-number': baSlovakPhoneNumberRegex,
  'ba-ratio': {
    validate: (value: string) => parseRatio(value).isValid,
  },
  'ba-ico': /^\d{6,8}$/,
  'ba-iban': {
    validate: (value: string) => {
      const electronicFormat = electronicFormatIBAN(value)
      if (!electronicFormat) {
        return false
      }
      return validateIBAN(electronicFormat).valid
    },
  },
}

export type BaAjvInputFormat = keyof typeof baAjvInputFormats

const baAjvOtherFormats = {
  'ba-time': {
    // https://stackoverflow.com/a/51177696
    validate: baTimeRegex,
    compare: compareBaTime,
  },
  'ba-file-uuid': {
    validate: validateBaFileUuid,
  },
}

export const baAjvFormats = {
  ...baAjvInputFormats,
  ...baAjvOtherFormats,
}
