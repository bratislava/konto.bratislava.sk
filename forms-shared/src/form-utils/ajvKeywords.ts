import 'json-schema';
import { JSONSchema7 } from 'json-schema'
declare module 'json-schema' {
  export interface JSONSchema7 {
    pospID?: string
    pospVersion?: string
    slug?: string
    hash?: string
    stepperTitle?: string
    file?: boolean
  }
}

export type BAJSONSchema7 = JSONSchema7

export const baAjvKeywords = [
  // Root schema
  {
    keyword: 'pospID',
  },
  {
    keyword: 'pospVersion',
  },
  {
    keyword: 'slug',
  },
  // Step schema
  {
    keyword: 'hash',
  },
  {
    keyword: 'stepperTitle',
  },
  // File field schema
  {
    keyword: 'file',
  },
]
