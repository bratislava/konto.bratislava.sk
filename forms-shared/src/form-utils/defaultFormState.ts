import {
  Experimental_DefaultFormStateBehavior,
  GenericObjectType,
  getDefaultFormState,
  RJSFSchema,
  ValidatorType,
} from '@rjsf/utils'

import { baRjsfValidator } from './validators'

/**
 * Detects schema of `fileUpload` with `multiple: true`.
 *
 * e.g. `{ type: 'array', items: { type: 'string', file: true } }`
 */
export const isFileMultipleSchema = (schema: RJSFSchema) =>
  !Array.isArray(schema?.items) &&
  typeof schema?.items !== 'boolean' &&
  schema?.items?.file === true

/**
 * This is the most important setting for RJSF, make sure to pass it to all RJSF components / functions. Anytime user
 * open a form, to display proper fields the data are prefilled by the library.
 *
 * `arrayMinItems` strategy:
 *
 * Let's break it down for each field type (defined in /generator/functions.ts):
 *  - `arrayField` - this is the only field that we want to get populated with default data if it is required - e.g.
 *    when user opens a form that contains list of "Tax records" and at least one is required, we want to display empty
 *    "Tax record" (which is an object with another fields) initially for user not to click on "+ Add tax record" button
 *    unnecessarily.
 *  - `selectMultiple` & `checkboxGroup` - these fields would get prefilled with `requiredOnly` strategy, but RJSF
 *    implements a special `isMultiSelect` check that flags those to not prefill.
 *    https://github.com/rjsf-team/react-jsonschema-form/blob/294b9e3d37c96888a0e8bb3c68a5b2b1afd452bf/packages/utils/src/schema/getDefaultFormState.ts#L403
 *  - `fileUpload` with `multiple: true` - these fields would get prefilled with `requiredOnly` strategy and RJSF doesn't
 *    handle this case, therefore custom `computeSkipPopulate` is needed, in case this is not present, RJSF prefills it
 *    as `[null]` which is not correct and causes bugs.
 *    This needed to be implemented for this use case: https://github.com/rjsf-team/react-jsonschema-form/pull/4121
 *
 * Why other strategies are not suitable:
 *  - `populate: 'always'` - this would prefill even not required `arrayField` fields
 *  - `populate: 'never'` - this would prefill nothing
 *
 * If a new field type is added to the generator, this needs to be reflected here. See tests for expected behavior.
 *
 * `allOf` strategy:
 *
 * This was a bug in RJSF:
 * https://github.com/rjsf-team/react-jsonschema-form/issues/3892
 * https://github.com/rjsf-team/react-jsonschema-form/issues/3832
 *
 * This should be a default behavior, but the maintainers decided to keep it as an experimental feature as it would be
 * a breaking change.
 */
export const baDefaultFormStateBehavior: Experimental_DefaultFormStateBehavior = {
  arrayMinItems: {
    populate: 'requiredOnly',
    computeSkipPopulate: (validator, schema) => isFileMultipleSchema(schema),
  },
  allOf: 'populateDefaults',
}

export const baGetDefaultFormState = (
  schema: RJSFSchema,
  formData: GenericObjectType,
  rootSchema?: RJSFSchema,
  customValidator?: ValidatorType,
) =>
  getDefaultFormState(
    customValidator ?? baRjsfValidator,
    schema,
    formData,
    rootSchema,
    undefined,
    baDefaultFormStateBehavior,
  )
