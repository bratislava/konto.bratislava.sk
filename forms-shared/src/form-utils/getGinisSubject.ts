import { FormDefinitionSlovenskoSkGeneric } from '../definitions/formDefinitionTypes'
import { GenericObjectType } from '@rjsf/utils'

export const getGinisSubject = (
  formDefinition: FormDefinitionSlovenskoSkGeneric,
  formData: GenericObjectType | null,
) => {
  const extractGinisSubject = formDefinition.ginisAssignment.extractGinisSubject
  if (!extractGinisSubject) {
    return formDefinition.title
  }

  return extractGinisSubject(formData) ?? formDefinition.title
}
