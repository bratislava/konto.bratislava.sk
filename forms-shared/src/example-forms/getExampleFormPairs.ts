import { FormDefinition } from '../definitions/formDefinitionTypes'
import {
  getFormDefinitionBySlug,
  getFormDefinitionBySlugDev,
} from '../definitions/getFormDefinitionBySlug'
import flatten from 'lodash/flatten'
import { exampleDevForms, exampleForms } from './exampleForms'
import { ExampleForm } from './types'

type ExampleFormPair<FD = FormDefinition> = {
  formDefinition: FD
  exampleForm: ExampleForm
}

export function getExampleFormPairs<FD extends FormDefinition>({
  formDefinitionFilterFn = (formDefinition: FormDefinition): formDefinition is FD => true,
  includeDevForms = false,
}: {
  formDefinitionFilterFn?: (formDefinition: FormDefinition) => formDefinition is FD
  includeDevForms?: boolean
} = {}): ExampleFormPair<FD>[] {
  const getFormDefinitionBySlugLocal = includeDevForms
    ? getFormDefinitionBySlugDev
    : getFormDefinitionBySlug
  const exampleFormsLocal = includeDevForms ? { ...exampleForms, ...exampleDevForms } : exampleForms

  const slugs = Object.keys(exampleFormsLocal)
  const formDefinitions = slugs
    .map(getFormDefinitionBySlugLocal)
    .filter(
      (formDefinition) => formDefinition != null && formDefinitionFilterFn(formDefinition),
    ) as FD[]

  const withExampleForms = formDefinitions.map((formDefinition) => {
    const exampleFormsBySlug = exampleFormsLocal[formDefinition.slug]
    return exampleFormsBySlug.map((exampleForm) => ({ formDefinition, exampleForm }))
  })

  return flatten(withExampleForms)
}
