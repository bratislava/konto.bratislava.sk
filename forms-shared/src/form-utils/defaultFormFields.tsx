import { FieldProps } from '@rjsf/utils'
import { getDefaultRegistry } from '@rjsf/core'
import React, { ComponentType, useMemo } from 'react'
import { BAJSONSchema7 } from './ajvKeywords'
import { defaultUiSchema } from './formDefaults'
import { getObjectFieldInfo } from './getObjectFieldInfo'

const assertBaOrderInProperties = (
  entries: [string, BAJSONSchema7][],
): entries is [string, { baOrder: number }][] => {
  return entries.every(([key, value]) => {
    return typeof key === 'string' && typeof value === 'object' && typeof value.baOrder === 'number'
  })
}

const defaultRegistry = getDefaultRegistry()
const { SchemaField, ObjectField } = defaultRegistry.fields

/**
 * Extracts `baUiSchema` from the schema and uses it as `uiSchema`.
 *
 * The RJSF says that "The `SchemaFieldRender` component is the work-horse of react-jsonschema-form", this is the top-level
 * component that renders each subschema in the form. Therefore, it is possible to extract the UI schema from the schema
 * here and let the original `SchemaField` act as it would be originally there.
 */
const ExtractUiSchemaSchemaField = (props: FieldProps) => {
  const { isFormObject } = getObjectFieldInfo(props.idSchema)
  const baUiSchema = props.schema.baUiSchema
  const uiSchemaEmptyObject = props.uiSchema ? Object.keys(props.uiSchema).length === 0 : false

  // `uiSchema` should be present only in the root object (whether `defaultUiSchema` or empty)
  if (isFormObject && (props.uiSchema === defaultUiSchema || uiSchemaEmptyObject)) {
    return <SchemaField {...props} />
  }

  if (props.uiSchema) {
    throw new Error('uiSchema is not supported')
  }

  return <SchemaField {...props} uiSchema={baUiSchema} />
}

/**
 * Dynamically creates `ui:order` based on `baOrder` values in schema.
 *
 * The input schema of this function is resolved object schema with merged all of subschemas (see fastMergeAllOf).
 */
const OrderObjectField = (props: FieldProps) => {
  if (props.uiSchema?.['ui:order']) {
    throw new Error('ui:order should not be present in the original uiSchema')
  }

  const newUiSchema = useMemo(() => {
    const entries = Object.entries(props.schema.properties as Record<string, BAJSONSchema7>)
    if (!assertBaOrderInProperties(entries)) {
      throw new Error('baOrder should be present in all properties')
    }

    const entriesSorted = entries.toSorted(
      ([, child1], [, child2]) => child1.baOrder - child2.baOrder,
    )
    const uiOrder = entriesSorted.map(([key]) => key)

    return {
      ...props.uiSchema,
      'ui:order': uiOrder,
    }
  }, [props.uiSchema, props.schema.properties])

  return <ObjectField {...props} uiSchema={newUiSchema} />
}

export type DefaultFormFieldType = 'SchemaField' | 'ObjectField'

export const defaultFormFields = {
  SchemaField: ExtractUiSchemaSchemaField,
  ObjectField: OrderObjectField,
} satisfies Record<DefaultFormFieldType, ComponentType<FieldProps>>
