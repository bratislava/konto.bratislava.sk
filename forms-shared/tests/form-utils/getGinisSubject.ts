import { describe, expect, test } from 'vitest'
import { getExampleFormPairs } from '../../src/example-forms/getExampleFormPairs'
import { getGinisSubject } from '../../src/form-utils/getGinisSubject'
import {
  FormDefinitionSlovenskoSkGeneric,
  isSlovenskoSkGenericFormDefinition,
} from '../../src/definitions/formDefinitionTypes'

describe('getGinisSubject', () => {
  test('should return form title if extractGinisSubject is not provided', () => {
    const formDefinition = {
      title: 'Test Form Title',
      ginisAssignment: {},
    } as FormDefinitionSlovenskoSkGeneric
    const formData = {}

    const result = getGinisSubject(formDefinition, formData)
    expect(result).toBe('Test Form Title')
  })

  test('should return result of extractGinisSubject if provided and returns value', () => {
    const formDefinition = {
      title: 'Test Form Title',
      ginisAssignment: {
        extractGinisSubject: (formData) => formData.subject,
      },
    } as FormDefinitionSlovenskoSkGeneric
    const formData = {
      subject: 'Extracted Subject',
    }

    const result = getGinisSubject(formDefinition, formData)
    expect(result).toBe('Extracted Subject')
  })

  test('should return form title if extractGinisSubject returns undefined', () => {
    const formDefinition = {
      title: 'Test Form Title',
      ginisAssignment: {
        extractGinisSubject: (formData) => undefined,
      },
    } as FormDefinitionSlovenskoSkGeneric
    const formData = {}

    const result = getGinisSubject(formDefinition, formData)
    expect(result).toBe('Test Form Title')
  })

  getExampleFormPairs({
    formDefinitionFilterFn: isSlovenskoSkGenericFormDefinition,
  }).forEach(({ formDefinition, exampleForm }) => {
    test(`${exampleForm.name} GINIS subject should match snapshot`, () => {
      const result = getGinisSubject(formDefinition, exampleForm.formData)
      expect(result).toMatchSnapshot()
    })
  })
})
