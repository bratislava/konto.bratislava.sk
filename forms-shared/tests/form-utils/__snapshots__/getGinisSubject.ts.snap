// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getGinisSubject > komunitneZahradyExample GINIS subject should match snapshot 1`] = `"Komunitné záhrady"`;

exports[`getGinisSubject > oznamenieOPoplatkovejPovinnostiZaKomunalneOdpadyExample GINIS subject should match snapshot 1`] = `"Oznámenie o poplatkovej povinnosti za komunálne odpady"`;

exports[`getGinisSubject > predzahradkyExample GINIS subject should match snapshot 1`] = `"Predzáhradky"`;

exports[`getGinisSubject > stanoviskoKInvesticnemuZameruExample GINIS subject should match snapshot 1`] = `"e-SIZ Dunajská Nový Bytový Dom, p.č. 56789 kú Karlova Ves, Dúbravka"`;

exports[`getGinisSubject > zavazneStanoviskoKInvesticnejCinnostiExample GINIS subject should match snapshot 1`] = `"e-ZST Dunajská Nový Bytový Dom, p.č. 56789 kú Karlova Ves, Dúbravka"`;

exports[`getGinisSubject > ziadostONajomBytuExample GINIS subject should match snapshot 1`] = `"Žiadosť o nájom bytu"`;

exports[`getGinisSubject > ziadostOSlobodnyPristupKInformaciamExample GINIS subject should match snapshot 1`] = `"Žiadosť o sprístupnenie informácií podľa zákona č. 211/2000 Z.z."`;
